### PR TITLE
[hotfix][core] Load balancing spillback feature flag

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This is a bandaid for #14420 which allows us to revert to the old scheduling behavior, but leaves the uniform random spillback policy feature flagged (on by default). It can be turned off via the environment variable `RAY_SCHEDULER_LOADBALANCE_SPILLBACK=0`. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
cc @stephanie-wang @rkooo567

File: python/ray/tests/test_multi_node_3.py [hunk 1]
```diff
@@ -392,52 +394,65 @@ def remote_print(s, file=None):
 def test_spillback_distribution(ray_start_cluster):
```
Comment (wuisawesome) [inline python/ray/tests/test_multi_node_3.py:394]: I changed the test because it always passed, even when we were deterministically spilling back.

**Comment by rkooo567** [comment]
Hmm why do we need bandaid pr? We have 3+ weeks until the next release.

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -107,6 +107,10 @@ RAY_CONFIG(bool, new_scheduler_enabled,
            getenv("RAY_ENABLE_NEW_SCHEDULER") == nullptr ||
                getenv("RAY_ENABLE_NEW_SCHEDULER") == std::string("1"))
 
+RAY_CONFIG(bool, scheduler_loadbalance_spillback,
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:110]: Can you keep this on by default to preserve existing behavior?

**Review by ericl — APPROVED** [review]
LGTM. Let's merge so we can easily experiment with the flag. @rkooo567 we should encourage feature flags like this, it makes it much easier to diagnose and revert issues without needing full recompiles.

**Comment by rkooo567** [comment]
I see. I also prefer to turn this option by default. Lgtm

**Comment by wuisawesome** [comment]
updated. will make sure the new unit test passes in ci then merge.

**Comment by wuisawesome** [comment]
> Hmm why do we need bandaid pr?

I'm actually ok with not merging this for now, but having CI build the wheels is really useful for investigating

**Comment by wuisawesome** [comment]
PR seems to work. Here are some stats:

Without loadbalancing (implies scheduling 5k tasks takes 70s): 
```
49
min: 1
max: 144
Time = 190.63857786800008
```

With loadbalancing (implies scheduling 5k tasks takes 170s):
```
64
min: 42
max: 87
Time = 294.70859461400005
```

For reference, 1.2.0 is (implies 5k tasks take 64s): 
```
42
min: 1
max: 144
Time = 184.7384821020005
```

**Comment by wuisawesome** [comment]
test script:

```
import ray
print(ray.__version__)

ray.init(address="auto")
from time import sleep, perf_counter
from tqdm import trange

@ray.remote(num_cpus=0.25)
def foo():
    sleep(120)
    return ray.worker.global_worker.current_node_id

start = perf_counter()
refs = [foo.remote() for _ in range(5000)]

remaining = refs
for _ in trange(5000):
    _, remaining = ray.wait(refs)
nodes = ray.get(refs)
end = perf_counter()

from collections import Counter
counter = Counter(nodes)

del counter[ray.worker.global_worker.current_node_id]

print(len(counter))
print(f"min: {min(counter.values())}")
print(f"max: {max(counter.values())}")

print(f"Time = {end - start}")

ray.shutdown()
```

**Comment by wuisawesome** [comment]
cluster setup:
```
# Experimental: an example of configuring a mixed-node-type cluster.
cluster_name: alex-status
max_workers: 999

# Cloud-provider specific configuration.
provider:
    type: aws
    region: us-west-2
    availability_zone: us-west-2a, us-west-2b, us-west-2c, us-west-2d
    cache_stopped_nodes: False

# Tell the autoscaler the allowed node types and the resources they provide.
# The key is the name of the node type, which is just for debugging purposes.
# The node config specifies the launch config and physical instance type.
available_node_types:
    head_node:
        node_config:
            InstanceType: r5dn.24xlarge
        min_workers: 0
        max_workers: 0
        resources: {}
    worker:
        node_config:
            InstanceType: c5.9xlarge
        resources: {}
        min_workers: 64
        max_workers: 64

# Specify the node type of the head node (as configured above).
head_node_type: head_node

# The default settings for the head node. This will be merged with the per-node
# type configs given above.
head_node:
    ImageId: latest_dlami

# The default settings for worker nodes. This will be merged with the per-node
# type configs given above.
worker_nodes:
    ImageId: latest_dlami

setup_commands:
    # - pip install -U https://ray-wheels.s3-us-west-2.amazonaws.com/spillback_feature_flag/1452ff37523766d93662d50d6869824547b48951/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
    - pip install -U ray==1.2.0
    - sudo bash -c 'rm -rf /etc/security/limits.d; echo "* soft nofile 65535" >> /etc/security/limits.conf; echo "* hard nofile 65535" >> /etc/security/limits.conf;'


head_start_ray_commands:
    - ray stop
    - export RAY_SCHEDULER_LOADBALANCE_SPILLBACK=1; ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml

```

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Hey should we just close it? I think we can make the hybrid implementation instead? (since there's a no blocker for that)

**Comment by wuisawesome** [comment]
wait the details of the hybrid approach aren't clear to me. What do you think they are?

**Comment by wuisawesome** [comment]
btw i think we should merge this anyways because this was a good learning experience that all scheduling spillback changes should be reversible

**Comment by rkooo567** [comment]
We run sequential scheduling, and if the object store memory or cpu usage is higher than 50%, prefer to spillback to other random nodes

**Comment by rkooo567** [comment]
But this means we will have 3 scheduling policies right? In this case, we should probably clean up code to support pluggable interface or sth

**Comment by wuisawesome** [comment]
Isn't that not going to work because we first pick the node to run a task on before we begin resolving dependencies? 

Unless you're suggesting we spillback even if the task _can_ run on the local node...

**Comment by rkooo567** [comment]
What about this? Do you think this won’t work? 

1. Check if locally schedulablr
2. If not, spillback (following eric’s protocol)
2. If yes, check load.
3. If load is high, see if there are low load nodes in other cluster.
4. If yes, spillback
5. If not, schedule locally. 

Other parts of the protocol could be the same?

**Comment by wuisawesome** [comment]
Isn't that going to result in bouncing around between nodes? I feel like we should merge this in the short term and then design a better medium spillback algorithm in the medium term, but our past 2 heuristics haven't been working out super well, so we probably shouldn't pump out a third one super quickly.

**Comment by rkooo567** [comment]
Hmm but the current sequential policy will likely to fail any of dask on ray user who will have some degree of multi node clusters though. 

Why do you think it is bouncing? Isn’t it technically same as our policy now? We just add one more resource constraint before spillback.

**Comment by ericl** [comment]
Let's merge this as a short term fix.

**Comment by ericl** [comment]
>            getenv("RAY_SCHEDULER_LOADBALANCE_SPILLBACK") != nullptr &&
               getenv("RAY_SCHEDULER_LOADBALANCE_SPILLBACK") != std::string("1"))

Hmm actually shouldn't the second part be `== std::string("1")`? Otherwise you have to turn it on by setting =0 which is weird.

**Pull request merged**